src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,551 @@
+from typing import List, Optional, Tuple
+
+import torch
+from torch import Tensor as T
+from torch import nn
+from torch.serialization import default_restore_location
+
+from .configuration_dpr import DprConfig
+from .file_utils import add_start_docstrings
+from .modeling_bert import BertConfig, BertModel
+from .modeling_utils import PreTrainedModel
+
+
+logger = logging.getLogger(__name__)
+
+DPR_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "dpr-base-uncased",
+]
+
+
+#############
+# BertEncoder
+#############
+
+
+class DprBertEncoder(BertModel):
+    def __init__(self, config, project_dim: int = 0):
+        BertModel.__init__(self, config)
+        assert config.hidden_size > 0, "Encoder hidden_size can't be zero"
+        self.encode_proj = nn.Linear(config.hidden_size, project_dim) if project_dim != 0 else None
+        self.init_weights()
+
+    @classmethod
+    def init_encoder(cls, cfg_name: str, projection_dim: int = 0, dropout: float = 0.0, **kwargs) -> "DprBertEncoder":
+        cfg = BertConfig.from_pretrained(cfg_name)
+        if dropout != 0:
+            cfg.attention_probs_dropout_prob = dropout
+            cfg.hidden_dropout_prob = dropout
+        return cls(cfg, project_dim=projection_dim)
+
+    def forward(self, input_ids: T, token_type_ids: Optional[T], attention_mask: Optional[T]) -> Tuple[T, ...]:
+        if self.config.output_hidden_states:
+            sequence_output, pooled_output, hidden_states = super().forward(
+                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+            )
+        else:
+            hidden_states = None
+            sequence_output, pooled_output = super().forward(
+                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+            )
+        pooled_output = sequence_output[:, 0, :]
+        if self.encode_proj:
+            pooled_output = self.encode_proj(pooled_output)
+        return sequence_output, pooled_output, hidden_states
+
+    def get_out_size(self) -> int:
+        if self.encode_proj:
+            return self.encode_proj.out_features
+        return self.config.hidden_size
+
+
+###########
+# Reader
+###########
+
+
+class Reader(nn.Module):
+    def __init__(self, encoder: nn.Module, hidden_size):
+        super(Reader, self).__init__()
+        self.encoder = encoder
+        self.qa_outputs = nn.Linear(hidden_size, 2)
+        self.qa_classifier = nn.Linear(hidden_size, 1)
+
+    def forward(self, input_ids: T, attention_mask: T):
+        # notations: N - number of questions in a batch, M - number of passages per questions, L - sequence length
+        N, M, L = input_ids.size()
+        start_logits, end_logits, relevance_logits = self._forward(
+            input_ids.view(N * M, L), attention_mask.view(N * M, L)
+        )
+        return start_logits.view(N, M, L), end_logits.view(N, M, L), relevance_logits.view(N, M)
+
+    def _forward(self, input_ids, attention_mask):
```

Comment:
we usually don't have two `_forward()` functions => insert the code in the forward function or do you reuse this function somewhere else?